File: fabric/python/fabric/hints.py
from dataclasses import dataclass, field


@dataclass
class HandleHint:
    handle: str
    seq: int


@dataclass
class EpochResult:
    epoch_tip: int
    handles: list[HandleHint] = field(default_factory=list)


@dataclass
class SpaceHint:
    space: str
    epoch_tip: int
    seq: int
    delegate_seq: int


@dataclass
class HintsResponse:
    anchor_tip: int = 0
    spaces: list[SpaceHint] = field(default_factory=list)
    epochs: list[EpochResult] = field(default_factory=list)


def _hints_score(h: HintsResponse) -> int:
    score = 0
    for s in h.spaces:
        score += s.epoch_tip * 1000 + s.seq + s.delegate_seq
    for e in h.epochs:
        score += e.epoch_tip * 100
        for hh in e.handles:
            score += hh.seq
    return score


def CompareHints(a: HintsResponse, b: HintsResponse) -> int:
    """Returns >0 if a is fresher, <0 if b is fresher, 0 if equal."""
    sa, sb = _hints_score(a), _hints_score(b)
    if sa != sb:
        return 1 if sa > sb else -1
    if a.anchor_tip != b.anchor_tip:
        return 1 if a.anchor_tip > b.anchor_tip else -1
    return 0

File: fabric/kotlin/src/main/kotlin/org/spacesprotocol/fabric/Hints.kt
package org.spacesprotocol.fabric

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class HandleHint(
    val handle: String,
    val seq: Int,
)

@Serializable
data class EpochResult(
    @SerialName("epoch_tip") val epochTip: Int,
    val handles: List<HandleHint> = emptyList(),
)

@Serializable
data class SpaceHint(
    val space: String,
    @SerialName("epoch_tip") val epochTip: Int,
    val seq: Int,
    @SerialName("delegate_seq") val delegateSeq: Int,
)

@Serializable
data class HintsResponse(
    @SerialName("anchor_tip") val anchorTip: Int = 0,
    val spaces: List<SpaceHint> = emptyList(),
    val epochs: List<EpochResult> = emptyList(),
)

fun compareHints(a: HintsResponse, b: HintsResponse): Int {
    val scoreA = hintsScore(a)
    val scoreB = hintsScore(b)
    if (scoreA != scoreB) return scoreA.compareTo(scoreB)
    return a.anchorTip.compareTo(b.anchorTip)
}

private fun hintsScore(h: HintsResponse): Int {
    var score = 0
    for (s in h.spaces) {
        score += s.epochTip * 1000 + s.seq + s.delegateSeq
    }
    for (e in h.epochs) {
        score += e.epochTip * 100
        for (hh in e.handles) {
            score += hh.seq
        }
    }
    return score
}
